src/main/java/uk/gov/ons/ctp/response/notify/util/BitCalculator.java

```diff
@@ -0,0 +1,207 @@
+package uk.gov.ons.ctp.response.notify.util;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.math.BigInteger;
+
+/**
+ * This is a simple class to take a gov.notify key or other arbitrary string representing a number and a radix and
+ * return the number of bits of data in the key. This is primarily for working out the validity of notify keys.
+ * @author innesm
+ */
+public class BitCalculator {
+
+    private static final Pattern KEY_REGEXP = Pattern.compile("^([^-]+)-(.*)$");
+
+    private static final int MINIMUM_KEY_BITS = 256;
+    private static final int UUID_BITS = 128;
+    private static final int HEX_RADIX = 16;
+
+    /**
+     * A bag of stuff holding details of the supplied data.  Could probably boil this down to just bits and valid.
+     */
+    public static class KeyInfo {
+        public final String inputString;
+        public final BigInteger value;
+        public final String valueStr;
+        public final String keyName;
+        public final int bits;
+        public final int radix;
+        public final boolean valid;
+
+        /**
+         * Constructor for immutable object
+         * @param aInputString String to analyse
+         * @param aValue Value represented by input string
+         * @param aValueStr Value portion of input string
+         * @param aKeyName Name of a key
+         * @param aBits Number of bits represented by data in input string
+         * @param aRadix Radix of number in input string
+         * @param aValid Validity flag
+         */
+        public KeyInfo(
+                final String aInputString,
+                final BigInteger aValue,
+                final String aValueStr,
+                final String aKeyName,
+                final int aBits,
+                final int aRadix,
+                final boolean aValid) {
+            this.inputString = aInputString;
+            this.value = aValue;
+            this.valueStr = aValueStr;
+            this.keyName = aKeyName;
+            this.bits = aBits;
+            this.radix = aRadix;
+            this.valid = aValid;
+        }
+
+        @Override
+        public String toString() {
+            // Extremely verbose format mandated by checkstyle
```

Comment:
Really, then we should change the checkstyle configuration :(

Reply:
Done

Reply:
Done